blob/service.go

```diff
@@ -0,0 +1,308 @@
+}
+
+// getByCommitment retrieves the DAH row by row, fetching shares and constructing blobs in order to
+// compare Commitments. Retrieving is stopped once the requested blob/proof is found.
+func (s *Service) getByCommitment(
+	ctx context.Context,
+	height uint64,
+	nID namespace.ID,
+	commitment Commitment,
+) (*Blob, *Proof, error) {
+	log.Infow("requesting blob",
+		"height", height,
+		"nID", nID.String())
+
+	header, err := s.getByHeight(ctx, height)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var (
+		rawShares = make([]share.Share, 0)
+		proofs    = make(Proof, 0)
+		amount    int
+		blobShare *shares.Share
+	)
+
+	namespacedShares, err := s.sGetter.GetSharesByNamespace(ctx, header.DAH, nID)
+	if err != nil {
+		return nil, nil, err
+	}
+	for _, rowShares := range namespacedShares {
+		if len(rowShares.Shares) == 0 {
+			continue
+		}
+
+		rawShares = append(rawShares, rowShares.Shares...)
+		proofs = append(proofs, rowShares.Proof)
+
+		// reconstruct the `blobShare` from the first rawShare in range
+		// in order to get blob's length(first share will contain this info)
+		if blobShare == nil {
+			for i, sh := range rawShares {
+				bShare, err := shares.NewShare(sh)
+				if err != nil {
+					return nil, nil, err
+				}
+
+				// ensure that the first share is not a NamespacePaddingShare
+				// these shares are used to satisfy the non-interactive default rules
+				// and are not the part of the blob, so should be removed.
+				isPadding, err := bShare.IsPadding()
+				if err != nil {
+					return nil, nil, err
+				}
+
+				if isPadding {
+					continue
+				}
+
+				blobShare = &bShare
+				// save the length.
+				length, err := blobShare.SequenceLen()
+				if err != nil {
+					return nil, nil, err
+				}
+				amount = shares.SparseSharesNeeded(length)
+				rawShares = rawShares[i:]
+				break
+			}
+		}
+
+		// move to the next row if the blob is incomplete.
+		if amount > len(rawShares) {
+			continue
+		}
+
+		// reconstruct the Blob.
+		blob, err := sharesToBlobs(rawShares[:amount])
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// compare commitments.
+		if bytes.Equal(blob[0].Commitment(), commitment) {
+			return blob[0], &proofs, nil
+		}
+
+		// drop info of the checked blob
+		rawShares = rawShares[amount:]
+		blobShare = nil
+	}
+	return nil, nil, ErrBlobNotFound
+}
+
+// getBlobs retrieves the DAH and fetches all shares from the requested namespace.ID and converts
+// them to Blobs.
+func (s *Service) getBlobs(ctx context.Context, nID namespace.ID, root *share.Root) ([]*Blob, error) {
+	namespacedShares, err := s.sGetter.GetSharesByNamespace(ctx, root, nID)
+	if err != nil {
+		return nil, err
+	}
+	return sharesToBlobs(namespacedShares.Flatten())
+}
+
+// getByHeight returns ExtendedHeader by its height.
+// getByHeight ensures that the requested height is less or equal than the current network head.
+func (s *Service) getByHeight(ctx context.Context, height uint64) (*header.ExtendedHeader, error) {
```

Comment:
The thing that might be annoying about our general design of how we hand headers to the components that need them is that there can be a significant gap between store.Head and sync.Head. We cannot access headers between store.Head --> sync.Head even if there's pending headers verified and available for "use" in syncer. Maybe we should consider this issue further down the line because we basically could have (--time that it takes to sync to network head from LocalHead amount--) of headers in the pending that are just inaccessible to other modules bc we don't have a convenient way to access them right now

Reply:
we can have a fallback solution if the node hasn't synced to the netHead, and request a header from the trusted peer

Reply:
This will be fixed by removing the non-adjacency requirement in the header Store. Until then, it's totally normal to ask users to wait until the node is fully synced. (discussed with @renaynay in sync)